hledger-lib/Hledger/Data/Amount.hs

```diff
@@ -817,17 +823,25 @@ showMixedAmountB opts ma
 -- width. This does not honour displayOneLine: all amounts will be displayed as if
 -- displayOneLine were False.
 showMixedAmountLinesB :: AmountDisplayOpts -> MixedAmount -> [WideBuilder]
-showMixedAmountLinesB opts@AmountDisplayOpts{displayMaxWidth=mmax,displayMinWidth=mmin} ma =
-    map (adBuilder . pad) elided
+showMixedAmountLinesB opts ma = showMixedAmountLinesB' pad opts ma
   where
-    astrs = amtDisplayList (wbWidth sep) (showAmountB opts) . amounts $
+    pad amt width = amt{ adBuilder = WideBuilder (TB.fromText $ T.replicate w " ") w <> adBuilder amt }
+      where w = width - wbWidth (adBuilder amt)
+
+-- | Helper for showMixedAmountB to show a list of Amounts on multiple lines. This returns
+-- the list of WideBuilders: one for each Amount, and with elision but without
+-- padding.  This does not honour displayOneLine: all amounts will be displayed
+-- as if displayOneLine were False.
+showMixedAmountLinesB' :: (AmountDisplay -> Int -> AmountDisplay)
+                       -> AmountDisplayOpts -> MixedAmount -> [WideBuilder]
+showMixedAmountLinesB' pad opts@AmountDisplayOpts{displayMaxWidth=mmax, displayMinWidth=mmin} ma =
+    map (adBuilder . flip pad width) elided
+  where
+    astrs = amtDisplayList (wbWidth sep) (showAmountB opts) . orderedAmounts opts $
```

Comment:
I might be missing something, but `showMixedAmountLinesB'` seems to just be `showMixedAmountLinesB` with the option to not provide any padding. Couldn't this just be achieved with `showMixedAmountLinesB opts{displayMinWidth=Nothing}`, which would not require introducing a new rendering function?

Reply:
We pad to the maximum of `displayMinWidth` (if it exists) and the width of the maximum `WideBuilder`. https://github.com/simonmichael/hledger/pull/1626/files/90ed5aa2b43b23eaef87d2248036930b333723c0#diff-a305cb83032e81d09c773ed3ab1fe919f06f482406e4e48aa022e3c8a4c6e83aR843. I didn't spend that much time thinking about a simpler alternative though.

Reply:
I'd just like to avoid a proliferation of helper functions. I can go through and try to refactor after you submit, if you'd prefer.

Reply:
I can make it an option on the `AmountDisplayOpts` if that's the preferred method.

Reply:
Let me say the three use cases we currently have, to make sure I have my understanding correct. Is this the case?
1. `showMixedAmountLinesB opts{displayMinWidth=Nothing}`: pads each line to the maximum width of all amounts
2. `showMixedAmountLinesB opts{displayMinWidth=Just n}`: pads each line to the maximum width of all amounts, or n if that is greater
3. `showMixedAmountLinesB' const`: does no padding whatsoever

Reply:
Yes, I believe that is the current behavior

Reply:
I think we can get the exact same behaviour without a helper function. In particular, case 1 can also be achieved with `opts{displayMinWidth=Just 0}`. If we change `opts{displayMinWidth=Nothing}` to do no padding whatsoever, that would encompass all our needs. If we change the `displayMinWidth` field of `noColour` to `Just 0` then nothing would break.

Reply:
Sounds good. I looked at the other usages of `displayMinWidth` and they seem OK with this change and I implemented this.

One behavioral change is with the `StringFormatComponent` parser in case widths are not specified. https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/StringFormat.hs#L141. Now the default will not pad anything as well.

Reply:
Do any of the tests fail? The `renderRow` function will also pad everything out to the same width (previously we were doing this twice: a little inefficient), so I would expect that nothing would break with the change for `StringFormatComponent`

Reply:
The tests pass but there is a difference in output.

With `Just 0` (old behavior when `Just 0 == Nothing`)

```
$ stack exec -- hledger bal -f examples/bcexample.hledger --commodity-column assets.*etrade --format "%(total) %(account)"
5120.50 USD Assets:US:ETrade:Cash
70.00 GLD Assets:US:ETrade:GLD
17.00 ITOT Assets:US:ETrade:ITOT
36.00 VEA Assets:US:ETrade:VEA
294.00 VHT Assets:US:ETrade:VHT
--------------------
  70.00 GLD
 17.00 ITOT
5120.50 USD
  36.00 VEA
 294.00 VHT
```

With `Nothing` (new behavior)

```
5120.50 USD Assets:US:ETrade:Cash
70.00 GLD Assets:US:ETrade:GLD
17.00 ITOT Assets:US:ETrade:ITOT
36.00 VEA Assets:US:ETrade:VEA
294.00 VHT Assets:US:ETrade:VHT
--------------------
70.00 GLD
17.00 ITOT
5120.50 USD
36.00 VEA
294.00 VHT
```

Reply:
Hmm. Neither of those is ideal output, but we should definitely fix the left-aligned totals field.

It should be sufficient to change `mmin` here to `mmin <|> Just 0`.
https://github.com/simonmichael/hledger/blob/f12ff0df3a194b388a428cd9fd344dcde7925481/hledger/Hledger/Cli/Commands/Balance.hs#L463

Reply:
I would prefer to change it at parse site so it's not confusing if options are displayed at different places. e.g

```
-    return $ FormatField (isJust leftJustified) (parseDec minWidth) (parseDec maxWidth) f
+    return $ FormatField (isJust leftJustified) (parseMin minWidth) (parseDec maxWidth) f
     where
-      parseDec s = case s of
-        Just text -> Just m where ((m,_):_) = readDec text
-        _ -> Nothing
+      parseDec = fmap $ fst . head . readDec
+      parseMin = Just . fromMaybe 0 . parseDec
```

Thoughts?

Reply:
Yeah, that sounds good too. Maybe `parseDec minWidth` -> `parseDec minWidth <|> Just 0` would do the job.

Reply:
OTOH, this would make it so that on the command-line we wouldn't be able to specify 'no padding' directly (which is consistent with current behavior). That is, `%(total)` is equivalent to `%0(total)`.

Reply:
There is no 'regression' with that but it seems a bit unfortunate.